From 10fc94fd82d690acd8800130c39dbaf6aab33695 Mon Sep 17 00:00:00 2001
From: Eric Naim <dnaim@cachyos.org>
Date: Wed, 22 Jul 2026 15:46:01 +0700
Subject: [PATCH 119/243] list: Mark some functions as __always_inline

Eric Dumazet found that when the kernel is compiled with clang,
list_add() and friends are not inlined into calling functions. Mark
these functions as __always_inline to improve performance and have a
smaller binary as a result.

Link: https://lore.kernel.org/all/CANn89iJVQe=wedLheJmjZjOTJsWHijT0jZs=iRxKssJZbjAxHw@mail.gmail.com/
Signed-off-by: Eric Naim <dnaim@cachyos.org>
---
 include/linux/list.h | 14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)

diff --git a/include/linux/list.h b/include/linux/list.h
index 09d979976b3b..7dd5b81d9b3e 100644
--- a/include/linux/list.h
+++ b/include/linux/list.h
@@ -151,7 +151,7 @@ static inline bool __list_del_entry_valid(struct list_head *entry)
  * This is only for internal list manipulation where we know
  * the prev/next entries already!
  */
-static inline void __list_add(struct list_head *new,
+static __always_inline void __list_add(struct list_head *new,
 			      struct list_head *prev,
 			      struct list_head *next)
 {
@@ -172,7 +172,7 @@ static inline void __list_add(struct list_head *new,
  * Insert a new entry after the specified head.
  * This is good for implementing stacks.
  */
-static inline void list_add(struct list_head *new, struct list_head *head)
+static __always_inline void list_add(struct list_head *new, struct list_head *head)
 {
 	__list_add(new, head, head->next);
 }
@@ -186,7 +186,7 @@ static inline void list_add(struct list_head *new, struct list_head *head)
  * Insert a new entry before the specified head.
  * This is useful for implementing queues.
  */
-static inline void list_add_tail(struct list_head *new, struct list_head *head)
+static __always_inline void list_add_tail(struct list_head *new, struct list_head *head)
 {
 	__list_add(new, head->prev, head);
 }
@@ -221,7 +221,7 @@ static inline void list_add_tail_release(struct list_head *new,
  * This is only for internal list manipulation where we know
  * the prev/next entries already!
  */
-static inline void __list_del(struct list_head * prev, struct list_head * next)
+static __always_inline void __list_del(struct list_head * prev, struct list_head * next)
 {
 	next->prev = prev;
 	WRITE_ONCE(prev->next, next);
@@ -235,13 +235,13 @@ static inline void __list_del(struct list_head * prev, struct list_head * next)
  * WRITE_ONCE() overhead of a regular list_del_init(). The code that uses this
  * needs to check the node 'prev' pointer instead of calling list_empty().
  */
-static inline void __list_del_clearprev(struct list_head *entry)
+static __always_inline void __list_del_clearprev(struct list_head *entry)
 {
 	__list_del(entry->prev, entry->next);
 	entry->prev = NULL;
 }
 
-static inline void __list_del_entry(struct list_head *entry)
+static __always_inline void __list_del_entry(struct list_head *entry)
 {
 	if (!__list_del_entry_valid(entry))
 		return;
@@ -255,7 +255,7 @@ static inline void __list_del_entry(struct list_head *entry)
  * Note: list_empty() on entry does not return true after this, the entry is
  * in an undefined state.
  */
-static inline void list_del(struct list_head *entry)
+static __always_inline void list_del(struct list_head *entry)
 {
 	__list_del_entry(entry);
 	entry->next = LIST_POISON1;
-- 
2.55.0

